File: src/clj/visitor_pattern.clj
(ns visitor-pattern)

;; Multimethods
; assume we have 2 data types
(def c {:shape :circle :r 10})
(def r {:shape :rectangle :l 5 :w 10})

; to define a new method on these data types
(defmulti pdf-export :shape)

(defmethod pdf-export :circle [_]
  (println "Exporting circle to pdf.."))

(defmethod pdf-export :rectangle [_]
  (println "Exporting rectangle to pdf.."))

(defmethod pdf-export :default [shape]
  (throw (Exception. (str "unrecognized shape: " shape))))

; call method on each data
(pdf-export c)
(pdf-export r)

;; Protocols

File: src/java/VisitorPattern.java
abstract class Shape {
    abstract void accept(Visitor visitor);
}

class Circle extends Shape {
    @Override
    void accept(Visitor visitor) {
        visitor.visitCircle(this);
    }
}

class Rectangle extends Shape {
    @Override
    void accept(Visitor visitor) {
        visitor.visitRectangle(this);
    }
}

interface Visitor {
    void visitCircle(Circle c);

    void visitRectangle(Rectangle r);
}

class PdfExportVisitor implements Visitor {
    @Override
    public void visitCircle(Circle c) {
        System.out.println("Exporting circle to pdf..");
    }

    @Override
    public void visitRectangle(Rectangle r) {
        System.out.println("Exporting rectangle to pdf..");
    }
}

class VisitorPattern {
    public static void main(String[] args) {
        var circle = new Circle();
        var rectangle = new Rectangle();

        Visitor pdfExport = new PdfExportVisitor();

        circle.accept(pdfExport);
        rectangle.accept(pdfExport);
    }
}
